### ISSUE TITLE
Absolute path needed to render with kdenlive

### ISSUE THREAD

**Issue by RonOnGithub** [gyroflow/gyroflow-frei0r#3]
Hi,

I've only just learned about gyroflow, and it's looking really promising
thanks for all the work you've put in to this! :)

I've been getting it set up to use it with kdenlive, and the latest hurdle in
making that all work nice and painless seem to be that the plugin stores
the Project file as a relative path (relative to the &lt;mlt root="..."&gt; path).

So when I edit a new project, add the gyroflow effect to a clip and set the
project file, it all works great.

But if I save that project and then reopen kdenlive, or when I try to actually
render it, then gyroflow sees the relative path fragment as invalid, and I see
the same thing as in the third image shown here:
https://github.com/gyroflow/gyroflow-frei0r/issues/2

If I manually edit the kdenlive project file or mlt render script to make the
Project property an absolute path then everything works as expected.

I'm just not sure if that's something gyroflow is supposed to tell mlt with a
parameter in the effect xml file, or if it's supposed to handle it in the plugin
code by prepending the mlt root to a relative path before using it.

**Comment by AdrianEddy** [comment]
Hmm, that sounds like Kdenlive bug, beause in the plugin code, the path is not altered at all, never, it's just always treated as a string. The parameter is just string too, not path. The Path parameter type is created and handled by Kdenlive (based on the [xml](https://github.com/gyroflow/gyroflow-frei0r/blob/main/frei0r_gyroflow.xml))

**Comment by RonOnGithub** [comment]
Yeah, that was my first thought too, but after a bit of digging, the project file xml does seem to be quite consistent about making paths under the &lt;mlt root=''&gt; relative to it, and I found this report against mlt https://github.com/mltframework/shotcut/issues/710 which seems to indicate this is intentional.

I couldn't find another effect which uses a path to see how it handled this, but it would seem a little odd to have the path in the xml project file to the .gyroflow file be absolute when the path to the video source is relative.   I'm seeing:

  &lt;property name="Project"&gt;footage/PRO_VID_20240410_064418_018.gyroflow
  &lt;property name="kdenlive:originalurl"&gt;footage/PRO_VID_20240410_064418_018.mp4
  &lt;property name="kdenlive:original.resource"&gt;footage/PRO_VID_20240410_064418_018.mp4

Can we access the mlt root parameter in lib.rs to create a full path there, or do we really need kdenlive (actually mlt in this case) to pass us something more?

I can open an issue on the kdenlive side, but if we can fix it on this side by normalising the Project path if it is relative, that does seem like the most robust answer to my current (slim) understanding of what's happening in this code.

**Comment by AdrianEddy** [comment]
There's really nothing I can do on my side, the `frei0r` interface is dead simple and doesn't have any control over host application or anything host-specific. All that gyroflow plugin handles in lib.rs is a single String. The interpretation of it as a path is entirely on the host side and invisible to the frei0r plugin

**Comment by RonOnGithub** [comment]
Ok, thanks for clarifying that.  I started out looking at whether the parameter type for Project needed to be something other than 'url' in frei0r_gyroflow.xml, but then I wasn't sure if we also had access to more than just those parameter values.  I'll leave this open here for now so we've got somewhere to track this side of the discussion, and if we need to bounce some more questions off you, but I'll take it up on the kdenlive side and figure out what needs to be done there.  Thanks for the fast response!

**Comment by AdrianEddy** [comment]
In theory it would work if you could set the parameter in xml to string and just type the full path there manually, but it appears that there's [no such type](https://github.com/KDE/kdenlive/blob/master/data/effects/README.md) in the effect xml
